src/main/kotlin/com/terraformation/backend/customer/api/AdminController.kt

```diff
@@ -574,6 +590,63 @@ class AdminController(
     return testClock()
   }
 
+  @PostMapping("/createAppVersion")
+  fun createAppVersion(
+      @RequestParam appName: String,
+      @RequestParam platform: String,
+      @RequestParam minimumVersion: String,
+      @RequestParam recommendedVersion: String,
+      redirectAttributes: RedirectAttributes
+  ): String {
+    try {
+      appVersionStore.create(AppVersionsRow(appName, platform, minimumVersion, recommendedVersion))
+      redirectAttributes.successMessage = "App version created."
+    } catch (e: DuplicateKeyException) {
+      redirectAttributes.failureMessage = "An entry for that app name and platform already exists."
+    } catch (e: Exception) {
+      log.error("Failed to create app version", e)
+      redirectAttributes.failureMessage = "Create failed: ${e.message}"
+    }
+
+    return appVersions()
+  }
+
+  @PostMapping("/updateAppVersion")
+  fun updateAppVersion(
+      @RequestParam originalAppName: String,
+      @RequestParam originalPlatform: String,
+      @RequestParam appName: String,
+      @RequestParam platform: String,
+      @RequestParam minimumVersion: String,
+      @RequestParam recommendedVersion: String,
+      redirectAttributes: RedirectAttributes
+  ): String {
+    try {
+      appVersionStore.update(
+          AppVersionsRow(originalAppName, originalPlatform),
+          AppVersionsRow(appName, platform, minimumVersion, recommendedVersion))
+      redirectAttributes.successMessage = "App version updated."
+    } catch (e: DuplicateKeyException) {
+      redirectAttributes.failureMessage = "An entry for that app name and platform already exists."
```

Comment:
should be updatable yes?

Reply:
This is if the user edits the app name or platform and it collides with an existing entry for that combination. Added a comment to clarify.

Reply:
Got it, thanks. Ugh yes we don't want the app/platform names embedded in the url path.. otherwise we could accept the modifiable values alone in the payload.
Makes sense now.